[next] Tsl slice shader example (#1202)
threlte/threlte: apps/docs/src/content/learn/advanced/webgpu.mdx

---
category: Advanced
title: WebGPU and TSL
order: 1
---

<Tip type="warning">
  WebGPU support in Three.js is still in an early stage and is subject to change from Three.js
  version to version.
</Tip>
<Tip type="warning">
  We highly recommend targeting [version r171](https://github.com/mrdoob/three.js/releases/tag/r171)
  onwards because of potential [duplication and configuration
  issues](https://github.com/mrdoob/three.js/pull/29404).
</Tip>
## WebGPU

To use Three.js's WebGPU renderer, import it and then initialize it within your
`<Canvas>`'s `createRenderer` prop.

```svelte title="App.svelte" {4}+ {8-14}+
<script>
  import Scene from './Scene.svelte'
  import { Canvas } from '@threlte/core'

>   import * as THREE from 'three/webgpu'

Reviewer: Would recommend against wildcard imports in the docs to encourage best practices for tree shaking:
```suggestion
  import { WebGPURenderer } from 'three/webgpu'
```

</script>

<Canvas
  createRenderer={(canvas) => {

>     return new THREE.WebGPURenderer({

Reviewer: ```suggestion
    return new WebGPURenderer({
```

      canvas,
      antialias: true,
      forceWebGL: false
    })
  }}
>
  <Scene />
</Canvas>
```

<Tip type="note">

>   WebGPU is still an experimental browser api and at the time of writing has 73% global
>   [support](https://caniuse.com/?search=webgpu). For this reason, Three.js's webgpu renderer

Reviewer: ```suggestion
  WebGPU is still an experimental browser api and at the time of writing has [limited availability across major browsers](https://caniuse.com/?search=webgpu). For this reason, Three.js's webgpu renderer
```

Easier to keep updated 😂

  fallbacks to webgl when webgpu is not available.
</Tip>
<Tip type="tip">
  This same approach can be used to swap out the default renderer for any other custom renderer.
</Tip>

<Example path="renderers/WebGPU" />

<small>
  Adapted from [this Three.js
  example](https://threejs.org/examples/?q=webgpu#webgpu_performance_renderbundle).
</small>

The WebGPU renderer doesn't immediately render. If the renderer you provide needs to delay renderering, you
can defer rendering by initially setting the renderMode to `manual` like so:

```svelte title="App.svelte"
<script>
  import { Canvas, T } from '@threlte/core@next'
  import * as THREE from 'three/webgpu'
  let renderMode = $state('manual')
</script>

<Canvas
  {renderMode}
  createRenderer={(canvas) => {

>     const renderer = new THREE.WebGPURenderer({

Reviewer: ```suggestion
    const renderer = new WebGPURenderer({
```

will often throw an error when it detects this.

To circumvent this issue, the following can be added to your Vite config.

```js
// vite.config.js
optimizeDeps: {
  esbuildOptions: {
    target: 'esnext'
  }
},
build: {
  target: 'esnext'
}
```

Alternatively,
[`vite-plugin-top-level-await`](https://github.com/Menci/vite-plugin-top-level-await)
can be used, although less success has been reported with this method.

## TSL

A question that comes up often in Three.js development is "How do I extend Three.js's materials?".
External libraries such as [three-custom-shader-material](https://www.npmjs.com/package/three-custom-shader-material)
use a find and replace solution to get this job done. Three.js has identified
that it's not an ideal solution and recommends using the [Three.js Shading Language](https://github.com/mrdoob/three.js/wiki/Three.js-Shading-Language) or TSL for short.

The example below is an adaptation of [this](https://threejs.org/examples/?q=tsl#webgpu_tsl_angular_slicing)
Three.js example. There are many more [TSL examples](https://threejs.org/examples/?q=tsl)
within Three.js that you can use or adapt for your project.

<Example path="shaders/slice/main" />

> ### Extend the catalogue

Reviewer: Even though this is convenient it still has big implications for bundle size. I'd recommend warning about this caveat and [also providing the alternative method that verekia came up with](https://github.com/verekia/three-gpu-ecosystem-tests/blob/main/vite-ts-threlte8/src/App.svelte) (which I'd use - and probably always recommend for stuff that's going into production considering how much unused code is imported with this method)

Reviewer 2: Hmm I think I'd like to talk about both. Maybe start with the easy Laissez-faire option of going wholesale and then talking about caveats in a sentence or two afterwards. It's a good point and grisch said verekia had tested it out too.

Reviewer: Yep! Completely agreed. I'd recommend suggesting both approaches and making explicit the consequences of each.

Reviewer 2: I put a lot of thought into structuring and ordering. I hope ya'll like it 😄

function.

```svelte title="App.svelte" {3-4}+ {6}+
<script>
  import Scene from './Scene.svelte'
  import { Canvas, extend } from '@threlte/core'
  import * as THREE from 'three/webgpu'

  extend(THREE)
</script>

<Canvas
  createRenderer={(canvas) => {
    return new THREE.WebGPURenderer({
      canvas,
      antialias: true,
      forceWebGL: false
    })
  }}
>
  <Scene />
</Canvas>
```

### Nodes

The [nodes](https://github.com/mrdoob/three.js/wiki/Three.js-Shading-Language#nodematerial) can be directly assigned like any other prop on the `<T>` component.

```svelte
<T.MeshPhysicalNodeMaterial
  outputNode={Fn(([arg1, arg2]) => {
    /* ... */
  })(arg1, arg2)}
  shadowNode={Fn(([arg1, arg2]) => {
    /* ... */
  })(arg2, arg2)}
/>
```

Node materials give you the ability to modify three's builtin materials.
In the sliced gear example, two nodes are modified; the `outputNode` and the `shadowNode`.
The `outputNode` is set up in such a way that it discards any fragments that are
outside the permitted `startAngle` and `arcAngle`. If a fragment is not discarded and
it is not front-facing, it is assigned the color in the `color` uniform. The material

> needs its `side` set to `THREE.DoubleSide` otherwise threejs will cull them out if they are facing away

Reviewer: ```suggestion
needs its `side` set to `THREE.DoubleSide` otherwise three.js will cull them out if they are facing away
```

To circumvent this issue, the following can be added to your Vite config.

```js
// vite.config.js
optimizeDeps: {
  esbuildOptions: {
    target: 'esnext'
  }
},
build: {
  target: 'esnext'
}
```

Alternatively,
[`vite-plugin-top-level-await`](https://github.com/Menci/vite-plugin-top-level-await)
can be used, although less success has been reported with this method.

## TSL

A question that comes up often in Three.js development is "How do I extend Three.js's materials?".
External libraries such as [three-custom-shader-material](https://www.npmjs.com/package/three-custom-shader-material)
use a find and replace solution to get this job done. Three.js has identified
that it's not an ideal solution and recommends using the [Three.js Shading Language](https://github.com/mrdoob/three.js/wiki/Three.js-Shading-Language) or TSL for short.

The example below is an adaptation of [this](https://threejs.org/examples/?q=tsl#webgpu_tsl_angular_slicing)
Three.js example. There are many more [TSL examples](https://threejs.org/examples/?q=tsl)
within Three.js that you can use or adapt for your project.

<Example path="shaders/slice/main" />

### Using the new classes

> By default `<T>` doesn't know about the new classes and will
> error on things like `<T.MeshPhysicalNodeMaterial />`. You have two options.

Reviewer: ```suggestion
`<T>` doesn't know about the new `WebGPURenderer` classes because they are part of a separate and new `three/webgpu` export, and creating components like `<T.MeshPhysicalNodeMaterial />` will throw errors. There are two options to fix this.
```

> 1. Give `<T>` all the new classes from `three/webgpu` with the [`extend`](/docs/reference/core/t#extending-the-default-component-catalogue)
>    function. Adding all the classes will increase the bundle size of your application
>    but is easy and works.

Reviewer: ```suggestion
1. Provide `<T>` with all the new classes from `three/webgpu` with the [`extend`](/docs/reference/core/t#extending-the-default-component-catalogue)
   function. It's important to note that using a wildcard import for adding all the classes will increase the bundle size of your application because both `three` and `three/webgpu` will be imported in a non-tree-shakable way. However, this method is easy and works.
```

**a)** use `<T>`'s [`is`](/docs/reference/core/t#property-is) prop with the webgpu
classes you initialize within your `<script></script>` tags

```svelte title="Scene.svelte" {3}+ {5}+ {10}+
<script>
  import { T } from '@threlte/core'
  import { MeshPhysicalNodeMaterial } from 'three/webgpu'

	const material = new MeshPhysicalNodeMaterial()
</script>

<T.Mesh>
	<T.BoxGeometry>
	<T is={material}>
</T.Mesh>
```

or **b)** [`extend`](/docs/reference/core/t#extending-the-default-component-catalogue)
`<T>` with those classes.

```svelte title="App.svelte" {3-4}+ {6}+
<script>
  import Scene from './Scene.svelte'
  import { Canvas, extend } from '@threlte/core'
  import { WebGPURenderer, MeshPhysicalNodeMaterial } from 'three/webgpu'

  extend({ MeshPhysicalNodeMaterial })
</script>

<Canvas
  createRenderer={(canvas) => {
    return new WebGPURenderer({
      canvas,
      antialias: true,
      forceWebGL: false
    })
  }}
>
  <Scene />
</Canvas>
```

Bundle size of your application will remain minimal but you'll have to keep things updated as you go.

<Tip type="tip">
  An easy option for projects is to start with option #1 and then transition to option #2 when

>   bundle size becomes an issue.

Reviewer: ```suggestion
  bundle size becomes an issue or when you need to ship to production.
```
